src/librustc_mir/interpret/place.rs

```diff
@@ -978,41 +978,40 @@ where
         let (mplace, size) = match place.place {
             Place::Local { frame, local } => {
                 match self.stack[frame].locals[local].access_mut()? {
-                    Ok(local_val) => {
+                    Ok(&mut local_val) => {
                         // We need to make an allocation.
-                        // FIXME: Consider not doing anything for a ZST, and just returning
-                        // a fake pointer?  Are we even called for ZST?
-
-                        // We cannot hold on to the reference `local_val` while allocating,
-                        // but we can hold on to the value in there.
-                        let old_val =
-                            if let LocalValue::Live(Operand::Immediate(value)) = *local_val {
-                                Some(value)
-                            } else {
-                                None
-                            };
 
                         // We need the layout of the local.  We can NOT use the layout we got,
                         // that might e.g., be an inner field of a struct with `Scalar` layout,
                         // that has different alignment than the outer field.
-                        // We also need to support unsized types, and hence cannot use `allocate`.
                         let local_layout = self.layout_of_local(&self.stack[frame], local, None)?;
-                        let (size, align) = self.size_and_align_of(meta, local_layout)?
-                            .expect("Cannot allocate for non-dyn-sized type");
-                        let ptr = self.memory.allocate(size, align, MemoryKind::Stack);
-                        let mplace = MemPlace { ptr: ptr.into(), align, meta };
-                        if let Some(value) = old_val {
-                            // Preserve old value.
-                            // We don't have to validate as we can assume the local
-                            // was already valid for its type.
-                            let mplace = MPlaceTy { mplace, layout: local_layout };
-                            self.write_immediate_to_mplace_no_validate(value, mplace)?;
+
+                        // Don't do anything for a ZST, and just return a valid integer
```

Comment:
```suggestion
                        // Don't do anything for a ZST, and just return a valid integer.
```
Also if this is about consistency with codegen, there should be a comment explaining that. But wouldn't it be easier to just codegen all empty allocs to integer ptrs when generating LLVM IR? This feels like codegen-only concerns bleeding into CTFE. IMO something is off with special-casting ZST here.

Reply:
If we don't do it during CTFE, then `const _: () = assert!(&() as *const () as usize == 1);` would fail, but `assert!(&() as *const () as usize == 1);` would succeed at runtime. Or we'd have to start supporting some form of typed `force_ptr`.

Reply:
context: https://github.com/rust-lang/rust/pull/62487 and https://github.com/rust-lang/rust/pull/63635 (my original link was wrong as you correctly mentioned)

Reply:
Things like
```rust
assert!(&42i32 as *const _ as usize == 0x1234500);
```
also can succeed at run-time but be made to fail by const-prop or other optimizations. That is entirely okay. The allocator is non-deterministic.

Reply:
I adjusted the PR to not change any values anymore. It's now purely an ICE fix + the change to run `const_eval` instead of `const_eval_raw` when we want an `Operand`

Reply:
oh... but this may break `static FOO: () = FOO;` I'll add a test

Reply:
Nope, doesn't break it. All good.